### PR TITLE
👷 Add Python venv cache

### PR DESCRIPTION
**Pull request by Kludex**
Following the same idea as https://github.com/encode/httpx/pull/1680 and https://github.com/encode/uvicorn/pull/1095.

This PR aims to add a cache to the Python environment.

### CODE REVIEW

**Comment by Kludex** [comment]
I'll put on "hold" until the `httpx` PR is merged, I want to see if @tomchristie has other concerns.

File: requirements.txt [hunk 1]
```diff
@@ -1,31 +1,25 @@
-# Optionals
-graphene; python_version<'3.10'
-itsdangerous
-jinja2
-python-multipart
-pyyaml
-requests
+-e .[full]
```
Comment (JayH5) [inline requirements.txt:1]: Do we really want to do `[full]`? These dependencies then won't be pinned and so will only be updated implicitly when the cache is invalidated.

File: .github/workflows/test-suite.yml [hunk 2]
```diff
@@ -21,13 +21,25 @@ jobs:
       - uses: "actions/setup-python@v2"
         with:
           python-version: "${{ matrix.python-version }}"
+
+      - uses: actions/cache@v2.1.6
```
Comment (JayH5) [inline .github/workflows/test-suite.yml:25]: What's the reason for asking for such a specific version?

**Review by JayH5 — COMMENTED** [review]
1. The implication here is that we now pin our dev dependencies and somebody updates them every once in a while? (Seems we're not so keen on dependabot because it creates too many PRs)
2. Before/after "benchmarks" would be appreciated 🙏  although I understand you're waiting for this to move further along with httpx for now, so no rush.

**Comment by JayH5** [comment]
I think overall I don't really have a problem with the build time as things are (~3min or less) so I'm not sure this is worth the extra complexity?

**Comment by lovelydinosaur** [comment]
> I'll put on "hold" until the `httpx` PR is merged, I want to see if @tomchristie has other concerns.

Echoing https://github.com/encode/uvicorn/pull/1095#issuecomment-869564569 Agreed. Let's take things one step at a time. 😀

> I think overall I don't really have a problem with the build time as things are (~3min or less) so I'm not sure this is worth the extra complexity?

I do also share this outlook somewhat. Not definitively, but tentatively.

**Comment by Kludex** [comment]
I'm closing this, as we still have to discuss on HTTPX.

**Pull request closed without merging**